# SNP-calling-pipeline

## Usage:

call_snps_pipeline_v9.pl -r <reference_genome> [-nosnp T/F default F][-o <oligos_file>] [-R <parent/reference fastq prefix if any>] -S [skip mappings, just do SNP calls, defaulf = F]  -D T/F [delete temp files default F] -f <file with other fastq file prefixes> [-c chromosome_ID] -A <reference annotation [RH88/GT1/ME49/Pf3D7/ATH/BESB1 other path to file]>
  
  Required parameters:
  
  -r <reference_genome> Add reference genome in fasta format
  
  -f <Prefix file> File containing fastq file prefixes, one prefix per line.
  
  Optional parameters:
  
  -nosnp [T/F, default F] Just run read mapping withou calling SNPs/InDels
  
  -o <oligos_file> Fasta file 
  
  -R <parent/reference fastq prefix if any> Use specified sample as the reference strain
  
  -S [T/F, default F] Skip mappings, just do SNP calls.
  
  -D [T/F, dafault F] Delete intermediate files generated by the pipeline.
  
  -c chromosome_ID
  
  -A <code for reference annotation database> [RH88|GT1|ME49|Pf3D7|ATH|BESB1|other path to file]
  
  Figure 1: SNP calling pipeline scheme.
  ![SNP_pipeline_scheme-9](https://user-images.githubusercontent.com/76788039/128446059-be2165d0-f347-4b40-b5d8-ed5a418877a7.png)


  
  
  
